### PR TITLE
Some galata improvements

### PR DESCRIPTION
**Pull request by fcollonval**
## References

Improve galata

Fixes https://github.com/jupyterlab/jupyterlab/issues/15372

## Code changes

- Extract better debugger helpers from https://github.com/jupyterlab/jupyterlab/pull/15156
- Better and faster `setCell`
- Hide the `test-*` folders in filebrowser for documentation snapshots. We start to see discrepancy because some times the test are executed with 2 workers instead of 1.
- Add new API to get locator instead of elementHandle

## User-facing changes

None

## Backwards-incompatible changes

None

### CODE REVIEW

**Comment by fcollonval** [comment]
Mike if you come by, I'm looking at this one right now.

**Comment by krassowski** [comment]
Thanks for the heads up!

**Comment by brichet** [comment]
@fcollonval what is the status of this PR ? 
Do you want to provide a copy of all the helpers returning handle, but returning a locator ? It would be great if this were the case, I can probably help on it if necessary.

**Comment by fcollonval** [comment]
> @fcollonval what is the status of this PR ? 
> Do you want to provide a copy of all the helpers returning handle, but returning a locator ? It would be great if this were the case, I can probably help on it if necessary.

Hey @brichet yes the main idea is to move to locators. Unfortunately lots of tests are still failing. If you have some cycles to look at the remaining failures your help will be much appreciated.

**Comment by brichet** [comment]
@fcollonval, I partially restored the previous logic to set cell mode.
Indeed double clicking set the edit mode for any cell type, but it can also select some text.
I think some tests were failing because of it.
Let me know if you think it is important to change this logic.

**Comment by brichet** [comment]
please update documentation snapshots

**Comment by brichet** [comment]
please update documentation snapshots

**Comment by brichet** [comment]
@fcollonval I think this PR is ready now. The remaining failures should not be related to this PR.
It removes all the use of element handles in the and `waitForSelector` (also deprecated) in the tests.
It adds new API in helpers to use locators instead of element handles, but keep the existing API.

We should add a deprecated comment on the previous API, to be able to remove it on the next major release of Galata.

**Comment by krassowski** [comment]
Sorry about the conflicts! I see that this PR adds many `*Locator` methods; there are two kinds of these:
- (a) `getCell(index: number): ElementHandle` now gets `getCellLocator(index: number): Locator`
- (b) `waitForTransition(element: ElementHandle<Element> | string): Promise<void>` now also gets `waitForLocatorTransition(element: Locator | string): Promise<void>` sibling

Both make tests a bit harder to read do to increased verbosity.  My thinking is that (a) is fine but (b) is not necessary. Instead we could have:

```js
waitForTransition(element: ElementHandle<Element> | Locator | string): Promise<void>
```

and deprecate passing `ElementHandle<Element>`. This would also reduce the delta of this PR. Thoughts?

**Comment by brichet** [comment]
Thanks for the comment @krassowski 

> My thinking is that (a) is fine but (b) is not necessary.

I agree, but I didn't find a way to share some code in the functions, while avoiding the use of `ElementHandle` if a `Locator` is provided. I'll try again, otherwise I'll duplicate the code.


> This would also reduce the delta of this PR.

Yes, I think 4 functions are impacted.

**Comment by brichet** [comment]
> Both make tests a bit harder to read do to increased verbosity. My thinking is that (a) is fine but (b) is not necessary. Instead we could have:
> 
> ```js
> waitForTransition(element: ElementHandle<Element> | Locator | string): Promise<void>
> ```

I updated it in https://github.com/jupyterlab/jupyterlab/pull/15375/files#diff-280c4ca697fcec0a3945f7919e10481049453ecbe9637fbfc3ce328c37f52dceR211, using `try catch` because I didn't found a way to discriminate `ElementHandle` from `Locator`, these are not classes but interfaces.
Let me know if you think of another implementation.

**Comment by brichet** [comment]
@fcollonval I rebased and opened the PR.

Let me know if you wanted to add some features in it.

**Comment by krassowski** [comment]
I think we should rebase and merge this one soon for it to avoid gathering more conflicts

**Comment by brichet** [comment]
> I think we should rebase and merge this one soon for it to avoid gathering more conflicts

I agree, I'll rebase it soon

**Comment by brichet** [comment]
@krassowski @fcollonval I rebased this PR, if you want to take a look

File: galata/src/helpers/notebook.ts [hunk 1]
```diff
@@ -599,25 +637,40 @@ export class NotebookHelper {
...
 
-    const isRenderedMarkdown = await cellEditor.evaluate(editor =>
-      editor.classList.contains('lm-mod-hidden')
-    );
+    const isRenderedMarkdown = (
+      await Utils.getLocatorClassList(cellEditor)
+    ).includes('lm-mod-hidden');
+    // const isRenderedMarkdown = await cellEditor.evaluate(editor =>
+    //   editor.classList.contains('lm-mod-hidden')
+    // );
```
Comment (krassowski) [inline galata/src/helpers/notebook.ts:671]: What is the story here?
```suggestion
```

**Review by krassowski — APPROVED** [review]
Thank you! From my point of view good to merge after cleaning up the commented code.

File: galata/src/helpers/notebook.ts (hunk 1 above)
Comment (brichet) [reply galata/src/helpers/notebook.ts]: Good catch, thanks

**Comment by brichet** [comment]
Thanks @krassowski for review.

Failed test is not related.

**Pull request merged**